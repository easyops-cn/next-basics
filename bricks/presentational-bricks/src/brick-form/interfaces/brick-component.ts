export interface BrickOptionProps {
  text: string;
  value?: string;
  id: string;
}
