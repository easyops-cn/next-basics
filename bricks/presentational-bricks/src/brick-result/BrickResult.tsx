import React, { CSSProperties } from "react";
import { Icon as LegacyIcon } from "@ant-design/compatible";
import { Result } from "antd";
import { ResultStatusType } from "antd/lib/result";
import { EmptyResult, EmptyResultStatus } from "@next-libs/basic-components";
import { getIllustration } from "@next-core/illustrations";
import { useCurrentTheme } from "@next-core/brick-kit";
import {
  BrickResultStatus,
  IllustrationsStatus,
} from "../interfaces/brick-result";

export interface IllustrationsConfig {
  imageStyle?: CSSProperties;
  name?: string;
  category?: string;
}

interface BrickResultProps {
  status: BrickResultStatus | EmptyResultStatus | IllustrationsStatus;
  title?: string;
  subTitle?: string;
  icon?: string;
  illustrationsConfig?: IllustrationsConfig;
}

export function BrickResult(props: BrickResultProps): React.ReactElement {
  const {
    illustrationsConfig: { name, category, imageStyle },
  } = props;
  const icon = props.icon ? <LegacyIcon type={props.icon} /> : "";
  const emptyResultStatus = Object.values(EmptyResultStatus);
  const theme = useCurrentTheme();
  const image = React.useMemo(() => {
    return getIllustration({ name, category, theme });
  }, [name, category, theme]);

  return emptyResultStatus.includes(props.status as EmptyResultStatus) ? (
    <Result
      title={props.title}
      subTitle={props.subTitle}
      icon={<EmptyResult status={props.status as EmptyResultStatus} />}
    >
      <slot id="content" name="content"></slot>
    </Result>
  ) : props.status === "illustrations" ? (
    <Result
      title={props.title}
      subTitle={props.subTitle}
      status={props.status as ResultStatusType}
      icon={<img src={image} style={imageStyle} />}
    >
      <slot id="content" name="content"></slot>
    </Result>
  ) : (
    <Result
      title={props.title}
      subTitle={props.subTitle}
      status={props.status as ResultStatusType}
      {...(icon ? { icon } : {})}
    >
      <slot id="content" name="content"></slot>
    </Result>
  );
}
