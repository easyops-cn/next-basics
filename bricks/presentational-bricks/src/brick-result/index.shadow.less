// --- 官方样式 ---
@import "~antd/lib/result/style/index.less";
@import "~antd/lib/empty/style/index.less";
// --- 自定义样式 ---
@import "~@next-core/custom-antd-styles/src/var.less";

.@{result-prefix-cls} {
  &-title {
    font-weight: var(--font-weight-500);
    color: var(--color-header-text);
  }
  &-content {
    font-size: var(--auxiliary-font-size);
    font-weight: var(--font-weight-400);
    color: var(--color-secondary-text);
    margin-top: 0;
    padding-top: 0;
    background-color: transparent;
  }
  &-subtitle {
    font-size: 16px;
    color: var(--color-secondary-text);
  }
}
