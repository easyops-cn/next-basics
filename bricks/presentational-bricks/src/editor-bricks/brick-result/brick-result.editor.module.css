.resultContainer {
  display: flex;
  flex-direction: column;
  justify-content: center;
  padding: var(--card-content-gap);
  gap: 20px;

  .icon {
    height: 130px;
    font-size: 100px;
    color: var(--editor-brick-stroke-color);
  }
  .contentWrapper {
    display: flex;
    flex-direction: column;
    align-items: center;

    .content {
      height: 20px;
      width: 220px;
      background-color: var(--editor-brick-skeleton-bg-light);
    }
  }
}
