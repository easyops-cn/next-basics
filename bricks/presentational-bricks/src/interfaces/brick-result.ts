import { CSSProperties } from "react";

export enum BrickResultStatus {
  Success = "success",
  Error = "error",
  Info = "info",
  Warning = "warning",
  E404 = "404",
  E403 = "403",
  E500 = "500",
}

export enum EmptyResultStatus {
  BrowserTooOld = "browser-too-old",
  Empty = "empty",
  NoData = "no-data",
  NoHistoryVersion = "no-history-version",
  NoVisitRecord = "no-visit-record",
  SearchEmpty = "search-empty",
  WelcomeToCreate = "welcome-to-create",
}

export type IllustrationsStatus = "illustrations";

export interface IllustrationsConfig {
  imageStyle?: CSSProperties;
  name?: string;
  category?: string;
}
