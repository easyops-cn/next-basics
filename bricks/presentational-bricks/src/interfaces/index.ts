export * from "./dropdown-select";
export * from "./common";
