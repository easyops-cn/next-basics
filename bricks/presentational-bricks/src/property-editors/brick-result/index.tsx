import type _React from "react";
import { getRuntime } from "@next-core/brick-kit";
import {
  brickResultSchema,
  IllustrationsStatusOptions,
  BrickResultStatusOptions,
  EmptyResultStatusOptions,
} from "./brickResult.schema";

function BrickResultComponentFactory(React: typeof _React) {
  return function BrickResultComponent(props: any): React.ReactElement {
    const {
      SchemaFieldComponent,
      formilySchemaFormatter,
      advancedMode,
      scope,
      form,
    } = props;

    React.useEffect(() => {
      form.setInitialValues({
        useNewIllustration: true,
        status: undefined,
        emptyResultSize: "middle",
      });
    }, [form]);

    return React.createElement(SchemaFieldComponent, {
      schema: formilySchemaFormatter(brickResultSchema as any, advancedMode!),
      scope: {
        ...scope,
        IllustrationsStatusOptions,
        BrickResultStatusOptions,
        EmptyResultStatusOptions,
      },
    });
  };
}

(getRuntime() as any).customEditors.define(
  "presentational-bricks.brick-result",
  BrickResultComponentFactory
);
