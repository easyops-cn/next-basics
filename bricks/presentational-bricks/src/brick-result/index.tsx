import React from "react";
import ReactDOM from "react-dom";
import { BrickWrapper, UpdatingElement, property } from "@next-core/brick-kit";
import { BrickResult } from "./BrickResult";
import style from "./index.shadow.less";
import {
  BrickResultStatus,
  EmptyResultStatus,
  IllustrationsStatus,
  IllustrationsConfig,
} from "../interfaces/brick-result";
import { IconSize } from "./components/IllustrationWrapper";
/**
 * @id presentational-bricks.brick-result
 * @name presentational-bricks.brick-result
 * @docKind brick
 * @description 结果页面，可配置其状态，标题，次要标题和自定义图标
 * @author ice
 * @slots
 * content: 提供 content 插槽
 * @history
 * 1.54.0:新增状态，`EmptyResultStatus`, 与规范对齐
 * @memo
 * @noInheritDoc
 */

export class BrickResultElement extends UpdatingElement {
  private _mountPoint: HTMLElement;

  /**
   * @required true
   * @default -
   * @description 结果的状态, 决定图标和颜色
   * @group basic
   */
  @property({ type: String })
  status: BrickResultStatus | EmptyResultStatus | IllustrationsStatus;

  /**
   * @required false
   * @description 主标题文字
   * @group basic
   */
  @property()
  customTitle: string;

  /**
   * @required false
   * @description 次标题文字
   * @group basic
   */
  @property()
  subTitle: string;

  /**
   * @required false
   * @default -
   * @description [自定义图标](https://ant.design/components/icon-cn/), 仅当 status 为 `BrickResultStatus` 时有效
   * @group ui
   */
  @property()
  icon: string;

  /**
   * @required false
   * @default {}
   * @description 自定义插画配置，size默认为middle，推荐使用size控制插画尺寸，可通过imageStyle覆盖size实现自定义大小（不推荐）
   * @group other
   */
  @property({ attribute: false })
  illustrationsConfig: IllustrationsConfig = {};

  /**
   * @default true
   * @required false
   * @description 在插画库的default分类下，使用新版本的图标替换default分类下图标
   * @group other
   */
  @property({ attribute: false })
  useNewIllustration = true;

  /**
   * @required false
   * @description status为EmptyResultStatus时，设置插画的尺寸
   * @group other
   */
  @property()
  emptyResultSize: IconSize;

  constructor() {
    super();

    const shadowRoot = this.attachShadow({ mode: "open" });
    const styleElement = document.createElement("style");
    styleElement.textContent = style;
    shadowRoot.appendChild(styleElement);
    this._mountPoint = document.createElement("div");
    shadowRoot.appendChild(this._mountPoint);
  }

  connectedCallback(): void {
    // istanbul ignore else
    if (!this.style.display) {
      this.style.display = "block";
    }
    this._render();
  }

  disconnectedCallback(): void {
    ReactDOM.unmountComponentAtNode(this);
  }

  protected _render(): void {
    // istanbul ignore else
    if (this.isConnected) {
      ReactDOM.render(
        <BrickWrapper>
          <BrickResult
            title={this.customTitle}
            status={this.status as BrickResultStatus}
            subTitle={this.subTitle}
            icon={this.icon}
            illustrationsConfig={this.illustrationsConfig}
            useNewIllustration={this.useNewIllustration}
            emptyResultSize={this.emptyResultSize}
          />
        </BrickWrapper>,
        this._mountPoint
      );
    }
  }
}

customElements.define("presentational-bricks.brick-result", BrickResultElement);
