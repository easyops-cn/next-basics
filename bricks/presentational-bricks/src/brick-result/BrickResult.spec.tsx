import React from "react";
import { shallow } from "enzyme";
import { BrickResult } from "./BrickResult";
import { BrickResultStatus } from "../interfaces/brick-result";
import { EmptyResultStatus } from "@next-libs/basic-components";
import { useFeatureFlags } from "@next-core/brick-kit";

jest.mock("@next-core/brick-kit");
(useFeatureFlags as jest.Mock).mockReturnValue([false]);

describe("BrickResult", () => {
  it("should work", () => {
    const wrapper = shallow(
      <BrickResult
        status={BrickResultStatus.Success}
        title="Success"
        illustrationsConfig={{}}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should work, with custom icon", () => {
    const wrapper = shallow(
      <BrickResult
        status={BrickResultStatus.Success}
        title="Success"
        icon="question"
        illustrationsConfig={{}}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("empty should work", () => {
    const wrapper = shallow(
      <BrickResult status={EmptyResultStatus.Empty} illustrationsConfig={{}} />
    );
    expect(wrapper.find("Result").prop("subTitle")).toBeUndefined();
  });
  it("should work with customize", () => {
    const wrapper = shallow(
      <BrickResult status={"illustrations"} illustrationsConfig={{}} />
    );
    expect(wrapper.find("Result").prop("status")).toBe("illustrations");
  });

  it("should work with featureFlags is true", () => {
    (useFeatureFlags as jest.Mock).mockReturnValue([true]);

    const wrapper = shallow(
      <BrickResult status={"illustrations"} illustrationsConfig={{}} />
    );
    expect(wrapper.find("Result").prop("status")).toBe("illustrations");
  });
});
