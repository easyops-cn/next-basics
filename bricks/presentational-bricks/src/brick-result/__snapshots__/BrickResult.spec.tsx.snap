// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrickResult should work 1`] = `
<Result
  status="success"
  title="Success"
>
  <slot
    id="content"
    name="content"
  />
</Result>
`;

exports[`BrickResult should work, with custom icon 1`] = `
<Result
  icon={
    <Icon
      type="question"
    />
  }
  status="success"
  title="Success"
>
  <slot
    id="content"
    name="content"
  />
</Result>
`;
